python/ray/data/datasource/tfrecords_datasource.py

```diff
@@ -361,6 +431,74 @@ def _read_records(
             raise RuntimeError(error_message) from e
 
 
+def _infer_schema_and_transform(dataset: "Dataset"):
+    list_sizes = dataset.aggregate(_MaxListSize(dataset.schema().names))
+
+    return dataset.map_batches(
+        _unwrap_single_value_lists,
+        fn_kwargs={"col_lengths": list_sizes["max_list_size"]},
+        batch_format="pyarrow",
+    )
+
+
+def _unwrap_single_value_lists(batch: pyarrow.Table, col_lengths: Dict[str, int]):
```

Comment:
Good catch, maybe in that case, it would be appropriate to catch the failure, and try applying the large_list -> list_ / large_binary -> binary casting.

Reply:
@scottjlee meaning that in this error scenario we will leave the user to cast the schema?

Reply:
I was thinking that Ray Data would automatically try the casting. Alternatively, we could add to the error message to indicate the user should modify the scheme themselves